Implement badge API for webviews and treeviews

matthewjamesadam opened the pull request:
This PR fixes https://github.com/microsoft/vscode/issues/62783
This PR implements the API proposal here: https://github.com/microsoft/vscode/pull/139225

I largely followed the existing patterns for how the TreeView and WebView processes state changes through the extension host into the main application.  I do have a few questions of the details of what I did, which I've added as PR comments.

I have only tested this manually, by modifying the TreeView and WebviewView sample extensions to provide badges, and running them in my own local build.  It works as expected.  I am not sure what the expectations are for writing automated tests for something like this, please let me know where I should add tests to a particular test suite.

I did notice one thing while working on this, which I think is worth discussing: in the WebView View badge case, I am not sure how badges can be provided after extension activation.  Webview Views specify a badge through `WebviewView.badge`, but you can only reference a `WebViewView` once it has been created (ie, once the user has opened up the Webview View, and `WebviewViewProvider.resolveWebviewView` has been called).  I'm wondering if this is insufficient though: imagine a TODO extension, like https://github.com/Gruntfuggly/todo-tree (they are one extension that wants to use this API).  If I used such an extension, I would expect that the badge would show up on the activation bar showing that I have TODOs as soon as I open up the workspace, even if I haven't yet opened up that extension's Webview View.

This is not a problem for TreeView badges, because an extension can call `createTreeView` to force the TreeView to be created when the extension is activated, and can provide badges at that point.

matthewjamesadam commented on src/vs/workbench/api/common/extHostTreeViews.ts:
```diff
@@ -406,6 +415,16 @@ class ExtHostTreeView<T> extends Disposable {
+
+	set badge(badge: vscode.ViewBadge | undefined) {
+		this._badge = badge;
+		this.proxy.$setBadge(this.viewId, badge);
```
I am passing the `vscode.ViewBadge` directly through to the proxy here (as the proxy's `IViewBadge` interface matches `vscode.ViewBadge`).  I would imagine that the proxy JSON-serializes this object and so this would work fine, but if it's better to make a new object here, or otherwise treat this data differently, let me know.

jrieken replied:
The better way is to copy all expected properties from `vscode.ViewBadge` into the target shape/object. In theory extensions can give you an object which more than `ViewBadge` containing lots of data or cycles

matthewjamesadam replied:
Done.

matthewjamesadam commented on src/vs/workbench/browser/parts/views/treeView.ts:
```diff
@@ -348,6 +350,29 @@ abstract class AbstractTreeView extends Disposable implements ITreeView {
+		if (this._badgeActivity) {
+			this._badgeActivity.dispose();
+			this._badgeActivity = undefined;
+		}
```
Here we track the last-set activity for each TreeView (and WebView, below), and whenever a new badge is set, we revoke the old activity and apply the new one.  This does work as expected but I'm not sure if there's a better way to do this.

alexr00 replied:
This is the way it's done in other places so it seems fine.

matthewjamesadam commented on src/vs/workbench/browser/parts/views/treeView.ts:
```diff
@@ -348,6 +350,29 @@ abstract class AbstractTreeView extends Disposable implements ITreeView {
+		if (badge) {
+			const activity = {
+				badge: new NumberBadge(badge.value, () => badge.tooltip)
+			};
```
I am only setting the badge property of `IActivity` here, there are also `clazz` and `priority` properties, I couldn't determine if they were meaningful for this use or not.

alexr00 replied:
It looks like most VS Code core badges use the default `priority`. The progress service uses 100 for the priority. Let's set the priority to 150 so that the progress service wins. I think most people will want to see the progress more than the number badge.

alexr00 replied:
For `clazz` it looks like that lets us specify additional css classes to add to the html element that ends up being added for the badge. We don't need to add this for now.

matthewjamesadam replied:
Done, I've updated the priority to `150` for both the tree and webviews.

mjbvz commented on src/vs/workbench/api/common/extHostTreeViews.ts:
```diff
@@ -406,6 +415,16 @@ class ExtHostTreeView<T> extends Disposable {
+		return this._badge;
+	}
+
+	set badge(badge: vscode.ViewBadge | undefined) {
```
We probably should skip the update here if the existing badge matches the current badge. This would be helpful in case a extension ends up mistakenly trying to update the badge multiple times in a loop

matthewjamesadam replied:
Done.

mjbvz reviewed (approved):
Webview parts of the change look good to me!

alexr00 commented on src/vs/workbench/browser/parts/views/treeView.ts:
```diff
@@ -348,6 +350,29 @@ abstract class AbstractTreeView extends Disposable implements ITreeView {
+	}
+
+	set badge(badge: IViewBadge | undefined) {
+
```
Before updating `this._badgeActivity` we should check to see if the properties of `badge` differ from `this._badge`.

matthewjamesadam replied:
Done.

alexr00 reviewed (changes requested):
Looks good, just a few minor changes!

matthewjamesadam commented:
Thanks for the review @alexr00 and @mjbvz , I believe I've addressed all of the comments.

I am still wondering about the question I added in the PR description, about how badges can be provided before the webview view is created -- I believe this is a fairly important use case.  Do either of you have any thoughts on this?

alexr00 commented on src/vs/workbench/api/common/extHostTreeViews.ts:
```diff
@@ -406,6 +415,21 @@ class ExtHostTreeView<T> extends Disposable {
+	}
+
+	set badge(badge: vscode.ViewBadge | undefined) {
+		if (this._badge?.value === badge?.value &&
```
We should check if proposed API is enabled here.

alexr00 reviewed (approved):
Everything's good from the tree view side. I'll wait for @mjbvz's input before merging.

matthewjamesadam commented:
I've cleaned up the conflicts this PR had with `main`.  @alexr00 I believe @mjbvz has already approved this PR here: https://github.com/microsoft/vscode/pull/144775#pullrequestreview-906563822.

@mjbvz I'd appreciate any thoughts you have about the webview view initialization issue I outlined in the PR description, as I believe it might make the badge feature less useful for webview views.

matthewjamesadam commented:
@alexr00 @mjbvz we are excited to get this into nightly builds, and I believe I've addressed all the code review issues.  Can we merge this in?

alexr00 commented:
Merged, thank you for the PR! @mjbvz we can continue to address the webview view questions during API review.

matthewjamesadam commented:
Thanks @alexr00 and @mjbvz  -- I have opened a follow-on Issue here to discuss the webview view launching problem: https://github.com/microsoft/vscode/issues/146330
